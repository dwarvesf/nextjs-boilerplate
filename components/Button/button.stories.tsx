import { storiesOf } from '@storybook/react'
import React from 'react'
import Button from '.'

storiesOf('components/Button', module).add('basic', () => (
  <div className="space-y-4">
    <div className="space-x-4 flex items-center">
      <Button>Click me</Button>
      <Button size="sm">Click me</Button>
    </div>
  </div>
))
