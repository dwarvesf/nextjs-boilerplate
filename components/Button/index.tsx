import React from 'react'

const Button: React.FC<any> = ({ children }) => {
  return (
    <button className="px-5 py-3 bg-yellow-600 text-white rounded">
      {children}
    </button>
  )
}

export default Button
