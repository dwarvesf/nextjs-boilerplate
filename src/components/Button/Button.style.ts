import { tv } from 'tailwind-variants'

const focusRing = 'focus:ring-2 focus:ring-offset-2 focus:outline-none'

const styles = tv({
  base: 'relative font-medium',
  variants: {
    fullWidth: {
      true: 'w-full flex',
    },
    loading: {
      true: 'cursor-default',
    },
    size: {
      lg: 'text-base',
      md: 'text-sm',
      sm: 'text-xs',
    },
    appearance: {
      primary: '',
      secondary: '',
      link: '',
      default: '',
    },
    disabled: {
      true: '',
    },
  },
  compoundVariants: [
    {
      appearance: 'primary',
      disabled: false,
      className: [
        'border-transparent text-white bg-pink-600 hover:bg-pink-700 focus:ring-pink-500',
        focusRing,
      ],
    },
    {
      appearance: 'primary',
      disabled: true,
      className: 'bg-gray-100 text-gray-400',
    },
    {
      appearance: 'secondary',
      disabled: false,
      className: [
        'border-transparent text-white bg-gray-500 hover:bg-gray-600 focus:ring-gray-400',
        focusRing,
      ],
    },
    {
      appearance: 'secondary',
      disabled: true,
      className: 'bg-gray-100 text-gray-400',
    },
    {
      appearance: 'link',
      disabled: false,
      className: [
        'text-pink-600 hover:text-pink-700 focus:ring-pink-500 rounded-sm',
        focusRing,
      ],
    },
    {
      appearance: 'link',
      disabled: true,
      className: 'text-gray-400',
    },
    {
      appearance: 'default',
      disabled: false,
      className: [
        'border-gray-300 text-gray-700 bg-white hover:bg-gray-50 focus:ring-pink-500',
        focusRing,
      ],
    },
    {
      appearance: 'default',
      disabled: true,
      className: 'bg-gray-100 text-gray-400',
    },
    {
      appearance: ['primary', 'secondary', 'default'],
      size: 'lg',
      className: 'py-4 px-6 px-8',
    },
    {
      appearance: ['primary', 'secondary', 'default'],
      size: 'md',
      className: 'px-5 py-3',
    },
    {
      appearance: ['primary', 'secondary', 'default'],
      size: 'sm',
      className: 'py-2 px-4 px-5',
    },
    {
      appearance: ['primary', 'secondary', 'default'],
      className: 'rounded-md shadow-sm border',
    },
  ],
  defaultVariants: {
    fullWidth: false,
    loading: false,
    disabled: false,
    size: 'md',
    appearance: 'default',
  },
})

export default styles
