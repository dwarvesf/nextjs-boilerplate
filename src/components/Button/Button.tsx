import React from 'react'
import cx from 'classnames'
import { BaseButton, BaseButtonProps } from 'components/BaseButton'
import { IconSpinner } from 'components/icons/components/IconSpinner'
import { forwardRefWithAs } from 'utils/forwardRefWithAs'
import { WithChildren } from 'types/common'
import { ButtonAppearance } from '../BaseButton/types'
import styles from './Button.style'

export interface ButtonProps extends WithChildren<BaseButtonProps> {
  appearance?: ButtonAppearance
  fullWidth?: boolean
  iconPosition?: 'left' | 'right'
  loading?: boolean
  disabled?: boolean
  Icon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>>
}

export const Button = forwardRefWithAs<'button', ButtonProps>(
  (
    {
      size = 'md',
      appearance = 'default',
      Icon,
      iconPosition = 'left',
      children: originChildren,
      loading = false,
      className,
      fullWidth = false,
      ...props
    },
    ref,
  ) => {
    let children = originChildren

    const baseClassName = styles({
      fullWidth,
      loading,
      disabled: props.disabled,
      appearance,
      size,
    })

    if (loading) {
      children = (
        <>
          <span className="absolute inset-0 flex items-center justify-center">
            <IconSpinner />
          </span>
          <span className="text-transparent">{children}</span>
        </>
      )
    } else if (Icon) {
      children = (
        <>
          {iconPosition === 'left' && <Icon className="mr-1.5" />}
          <span>{children}</span>
          {iconPosition === 'right' && <Icon className="ml-1.5" />}
        </>
      )
    }

    return (
      <BaseButton ref={ref} {...props} className={cx(className, baseClassName)}>
        {children}
      </BaseButton>
    )
  },
)
