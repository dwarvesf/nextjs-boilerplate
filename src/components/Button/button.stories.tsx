import { storiesOf } from '@storybook/react'
import { IconCheckCircleSolid } from 'components/icons/components/IconCheckCircleSolid'
import { IconCloseCircleSolid } from 'components/icons/components/IconCloseCircleSolid'
import { IconExlamationCirleSolid } from 'components/icons/components/IconExlamationCirleSolid'
import { IconInformationCircleSolid } from 'components/icons/components/IconInformationCircleSolid'
import React from 'react'
import { Button } from '.'

storiesOf('components/Button', module).add('default', () => (
  <div className="space-y-4">
    <div className="space-x-4 flex items-center">
      <Button size="sm">Click me</Button>
      <Button>Click me</Button>
      <Button size="lg">Click me</Button>
      <Button loading>Click me</Button>
      <Button disabled>Click me</Button>
    </div>

    <div className="space-x-4 flex items-center">
      <Button appearance="primary" size="sm">
        Click me
      </Button>
      <Button appearance="primary">Click me</Button>
      <Button appearance="primary" size="lg">
        Click me
      </Button>
      <Button appearance="primary" loading>
        Click me
      </Button>
      <Button appearance="primary" disabled>
        Click me
      </Button>
    </div>

    <div className="space-x-4 flex items-center">
      <Button appearance="secondary" size="sm">
        Click me
      </Button>
      <Button appearance="secondary">Click me</Button>
      <Button appearance="secondary" size="lg">
        Click me
      </Button>
      <Button appearance="secondary" loading>
        Click me
      </Button>
      <Button appearance="secondary" disabled>
        Click me
      </Button>
    </div>

    <div className="space-x-4 flex items-center">
      <Button appearance="link" size="sm">
        Click me
      </Button>
      <Button appearance="link">Click me</Button>
      <Button appearance="link" size="lg">
        Click me
      </Button>
      <Button appearance="link" loading>
        Click me
      </Button>
      <Button appearance="link" disabled>
        Click me
      </Button>
    </div>

    <div className="space-x-4 flex items-center">
      <Button size="sm" Icon={IconCheckCircleSolid}>
        Click me
      </Button>
      <Button appearance="primary" Icon={IconExlamationCirleSolid}>
        Click me
      </Button>
      <Button
        appearance="secondary"
        Icon={IconInformationCircleSolid}
        size="lg"
      >
        Click me
      </Button>
      <Button appearance="primary" Icon={IconCloseCircleSolid} disabled>
        Click me
      </Button>
    </div>
  </div>
))
