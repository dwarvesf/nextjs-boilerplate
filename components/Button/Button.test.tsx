/**
 * @jest-environment jsdom
 */
import React from 'react'
import { render } from '@testing-library/react'
import Button from '.'

describe('<Button/>', () => {
  it('Should match snapshot', () => {
    const { container } = render(<Button>Hello</Button>)

    expect(container.firstChild).toMatchSnapshot()
  })
})
