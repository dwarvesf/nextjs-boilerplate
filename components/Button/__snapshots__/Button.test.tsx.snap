// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Should match snapshot 1`] = `
<button
  class="relative text-sm px-5 py-3  font-medium rounded-md shadow-sm border border-gray-300 text-gray-700 bg-white hover:bg-gray-50 focus:ring-pink-500 focus:ring-2 focus:ring-offset-2 focus:outline-none inline-flex justify-center items-center text-center whitespace-nowrap leading-none focus:outline-none transition duration-200 cursor-pointer"
>
  Hello
</button>
`;
